internal/decode/decoder.go

```diff
@@ -166,22 +170,65 @@ func unmarshalAsYAMLOrSchema(data []byte, out interface{}) (bool, error) {
 	return false, nil
 }
 
-func compileSchemaFromData(data []byte, out interface{}) error {
-	compiled, serr := compiler.Compile(compiler.InputSchema{
+// compileSchemaFromData attempts to compile using the old DSL and the new composable DSL,
+// but prefers the new DSL.
+// It returns the errors returned by both compilations.
+func compileSchemaFromData(filename, schemaString string, out interface{}) error {
+	var (
+		standardCompileErr   error
+		standardCompiled     *compiler.CompiledSchema
+		composableCompiled   *composable.CompiledSchema
+		composableCompileErr error
+		vfile                validationfile.ValidationFile
+	)
+
+	vfile = *out.(*validationfile.ValidationFile)
+
+	inputSourceFolder := filepath.Dir(filename)
+	composableCompiled, composableCompileErr = composable.Compile(composable.InputSchema{
+		SchemaString: schemaString,
+	}, composable.AllowUnprefixedObjectType(), composable.SourceFolder(inputSourceFolder))
+
+	standardCompiled, standardCompileErr = compiler.Compile(compiler.InputSchema{
 		Source:       input.Source("schema"),
-		SchemaString: string(data),
+		SchemaString: schemaString,
 	}, compiler.AllowUnprefixedObjectType())
-	if serr != nil {
-		return serr
+
+	if standardCompileErr == nil {
+		vfile.Schema = blocks.ParsedSchema{
+			CompiledSchema: standardCompiled,
+			Schema:         schemaString,
+			SourcePosition: spiceerrors.SourcePosition{LineNumber: 1, ColumnPosition: 1},
+		}
 	}
 
-	// If that succeeds, return the compiled schema.
-	vfile := *out.(*validationfile.ValidationFile)
-	vfile.Schema = blocks.ParsedSchema{
-		CompiledSchema: compiled,
-		Schema:         string(data),
-		SourcePosition: spiceerrors.SourcePosition{LineNumber: 1, ColumnPosition: 1},
+	if composableCompileErr == nil {
+		oldDefinitions := make([]compiler.SchemaDefinition, 0, len(composableCompiled.OrderedDefinitions))
+		for _, definition := range composableCompiled.OrderedDefinitions {
+			oldDefinition, ok := definition.(compiler.SchemaDefinition)
+			if !ok {
+				return fmt.Errorf("could not convert definition to old schemadefinition: %v", oldDefinition)
+			}
+			oldDefinitions = append(oldDefinitions, oldDefinition)
+		}
+
+		compiledSchemaString, _, err := generator.GenerateSchema(composableCompiled.OrderedDefinitions)
+		if err != nil {
+			return fmt.Errorf("could not generate string schema: %w", err)
+		}
+		vfile.Schema = blocks.ParsedSchema{
+			CompiledSchema: &compiler.CompiledSchema{
+				ObjectDefinitions:  composableCompiled.ObjectDefinitions,
+				CaveatDefinitions:  composableCompiled.CaveatDefinitions,
+				OrderedDefinitions: oldDefinitions,
+			},
+			Schema:         compiledSchemaString,
+			SourcePosition: spiceerrors.SourcePosition{LineNumber: 1, ColumnPosition: 1},
+		}
```

Comment:
This feels like an interim solution to me - the shapes happen to match, but that may not continue to be the case as each compiler package evolves, and we'll have deferred this work until then.

Was that an intentional decision? I'm fine with an answer of "yes", especially given that when I tried to change the validationfile package to use an interface, I disappeared down a rabbit hole for a week.

If it was, can we add a comment to that effect?

Reply:
So, the reason i'm doing this is that the `validate` code uses `var parsed validationfile.ValidationFile` and I didn't want to change that. 

And so, I borrowed code from https://github.com/authzed/zed/blob/d3826bf5872bd3722fdf46b85b6ecc05c89d6365/internal/cmd/preview.go#L90-L101

I guess, looking more in detail, I only need the `compiledSchemaString` 🤔